static/js/src/advantage/manage-subscription.js

```diff
@@ -1,93 +1,259 @@
-        setTimeout(function () {
-          location.reload();
-        }, 2000);
+  return default_message;
+};
+
+const handleSuccess = (elements, code) => {
+  const message = getMessage(code, "Success!");
+
+  elements["button"]["html"].classList.remove("is-processing");
+  elements["button"]["html"].innerHTML = elements["button"]["text"];
+  elements["button"]["html"].disabled = false;
+  elements["success"].forEach((element) => {
+    element.classList.remove("u-hide");
+    element.querySelector(".p-notification__message").innerHTML = message;
+  });
+
+  setTimeout(() => {
+    location.reload();
+  }, 2000);
+};
+
+const handleError = (elements, code) => {
+  const message = getMessage(code, "Error!");
+
+  elements["button"]["html"].classList.remove("is-processing");
+  elements["button"]["html"].innerHTML = elements["button"]["text"];
+  elements["button"]["html"].disabled = false;
+  elements["caution"].forEach((element) => {
+    element.classList.remove("u-hide");
+    element.querySelector(".p-notification__message").innerHTML = message;
+  });
+};
+
+const authenticate_3ds = (invoice, elements) => {
+  stripe.confirmCardPayment(invoice.pi_secret).then(function (data) {
+    if (data.error) {
+      handleError(elements, "resizing_machines_fail");
+
+      return;
+    }
+
+    handleSuccess(elements, "resizing_machines_success");
+  });
+};
+
+// It take a time for the invoice to be attached to the payment.
+// We try 5 times each 2 seconds.
+const retryGetPurchase = (purchase_id, attempts_counter, elements) => {
+  let max_number_of_attempts = 5;
+  if (attempts_counter > max_number_of_attempts) {
+    handleError(elements, "resizing_machines_failed");
+  }
+
+  setTimeout(function () {
+    getPurchase(purchase_id).then((purchase) => {
+      if (purchase.status === "done") {
+        handleSuccess(elements, "resizing_machines_success");
+
+        return;
+      }
+
+      let invoice;
+
+      if (purchase.stripeInvoices && purchase.stripeInvoices.length > 0) {
+        invoice = purchase.stripeInvoices[0];
+      }
+
+      if (!invoice) {
+        attempts_counter++;
+        retryGetPurchase(purchase_id, attempts_counter);
+
+        return;
+      }
+
+      if (invoice.pi_status === "requires_payment_method") {
+        handleError(elements, "resizing_machines_fail");
+
+        return;
+      }
+
+      if (requiresAuthentication(invoice)) {
+        authenticate_3ds(invoice, elements);
       }
-    })
-    .catch((error) => {
-      console.error(error);
     });
-}
+  }, 2000);
+};
 
-function cancelSubscription(id, VPSize) {
-  const cancelSubscriptionButton = document.querySelector(
-    `#cancel-subscription--${id}[data-viewport="${VPSize}"]`
+const requiresAuthentication = (invoice) => {
+  if (invoice.pi_decline_code === "authentication_required") {
+    return true;
+  }
+
+  if (invoice.pi_status === "requires_action" && invoice.pi_secret) {
+    return true;
+  }
+
+  return false;
```

Comment:
```suggestion
return invoice.pi_decline_code === "authentication_required" || (invoice.pi_status === "requires_action" && invoice.pi_secret)
```

Since this is all we do I'm not sure we need a function here. Do you think writing the condition like that is not legible enough?

Reply:
I like this one. I do think we should keep the function.